src/trackinfoobject.cpp

```diff
@@ -865,96 +869,80 @@ bool TrackInfoObject::isDirty() {
 
 int TrackInfoObject::getRating() const {
     QMutexLocker lock(&m_qMutex);
-    return m_Rating;
+    return m_iRating;
 }
 
 void TrackInfoObject::setRating (int rating) {
     QMutexLocker lock(&m_qMutex);
-    if (rating != m_Rating) {
-        m_Rating = rating;
-        setDirty(true);
+    if (modifyFieldValue(&m_iRating, rating)) {
+        markDirtyAndUnlock(&lock);
     }
 }
 
-void TrackInfoObject::setKeys(Keys keys) {
+void TrackInfoObject::setKeys(const Keys& keys) {
     QMutexLocker lock(&m_qMutex);
-    setDirty(true);
+    setKeysAndUnlock(&lock, keys);
+}
+
+void TrackInfoObject::setKeysAndUnlock(QMutexLocker* pLock, const Keys& keys) {
     m_keys = keys;
-    // Might be INVALID. We don't care.
+    // New key might be INVALID. We don't care.
     mixxx::track::io::key::ChromaticKey newKey = m_keys.getGlobalKey();
-    lock.unlock();
+    markDirtyAndUnlock(pLock);
     emit(keyUpdated(KeyUtils::keyToNumericValue(newKey)));
     emit(keysUpdated());
 }
 
-const Keys& TrackInfoObject::getKeys() const {
+Keys TrackInfoObject::getKeys() const {
     QMutexLocker lock(&m_qMutex);
     return m_keys;
 }
 
-mixxx::track::io::key::ChromaticKey TrackInfoObject::getKey() const {
-    QMutexLocker lock(&m_qMutex);
-    if (!m_keys.isValid()) {
-        return mixxx::track::io::key::INVALID;
-    }
-    return m_keys.getGlobalKey();
-}
-
 void TrackInfoObject::setKey(mixxx::track::io::key::ChromaticKey key,
-                             mixxx::track::io::key::Source source) {
-    QMutexLocker lock(&m_qMutex);
-    bool dirty = false;
+                             mixxx::track::io::key::Source keySource) {
     if (key == mixxx::track::io::key::INVALID) {
-        m_keys = Keys();
-        dirty = true;
-    } else if (m_keys.getGlobalKey() != key) {
-        m_keys = KeyFactory::makeBasicKeys(key, source);
-    }
-
-    if (dirty) {
-        setDirty(true);
+        resetKeys();
+    } else {
+        Keys keys(KeyFactory::makeBasicKeys(key, keySource));
+        QMutexLocker lock(&m_qMutex);
+        if (m_keys.getGlobalKey() != key) {
+            setKeysAndUnlock(&lock, keys);
+        }
     }
-
-    // Might be INVALID. We don't care.
-    mixxx::track::io::key::ChromaticKey newKey = m_keys.getGlobalKey();
-    lock.unlock();
-    emit(keyUpdated(KeyUtils::keyToNumericValue(newKey)));
-    emit(keysUpdated());
 }
 
-void TrackInfoObject::setKeyText(QString key,
-                                 mixxx::track::io::key::Source source) {
```

Comment:
One of the issues that affected many setters was that signals were emitted 
in setDirty() while the TIO was still locked by a QMutexLock in the calling 
function. This could lead to deadlocks.

On 01/10/2016 09:22 PM, Daniel Schürmann wrote:
>
> In src/trackinfoobject.cpp 
> <https://github.com/mixxxdj/mixxx/pull/852#discussion_r49280912>:
>
> >  }
> >
> > -void TrackInfoObject::setKeyText(QString key,
> > -                                 mixxx::track::io::key::Source source) {
>
> The diff is hard to read. What was the issue with setKeyText?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/852/files#r49280912>.
>